### DISCUSSION TITLE
Basic question about writing effects in terms of other effects

### DISCUSSION THREAD

**Discussion by sullyj3** [haskell-effectful/effectful#177] [Q&A] [answered]
My question is, must the "higher level" effect always be dynamically dispatched?

To make it concrete, suppose I wanted to write an effect `Hello` with a single nullary operation `hello` which dumps a text file containing "hello world". If I want to implement this in terms of `FileSystem`, which is statically dispatched, must `Hello` be dynamically dispatched? (I'm hazy on the details of how to do that actually, I think you use `reinterpret`?). It feels like it ought to be able to be static, since the implementation never changes.

**Comment by arybczak** [comment, marked as answer]
> My question is, must the "higher level" effect always be dynamically dispatched?

Yes. If you want to depend on an existing effect, it has to be dynamically dispatched. Static effects are primitive in a sense as they can't use other effects.

But considering that the `FileSystem` effect itself is static and just a thin wrapper for stuff from `base`/`bytestring`, you can just make `hello` use [hPutStrLn](https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO.html#v:hPutStrLn) (or `writeFile`) directly.

**Reply by sullyj3** [reply]
Ok, thanks. Yeah, `Hello`/`FileSystem` was just an example, in the actual case I'm interested in the lower level effect has a bunch of functionality I don't want to reimplement. I'll just make the higher level one dynamic.
